resolver/resolver.cc

```diff
@@ -1317,16 +1374,60 @@ class ResolveConstantsWalk {
     }
 
 public:
-    ResolveConstantsWalk() : nesting_(nullptr) {}
+    ResolveConstantsWalk() : nesting_(nullptr), loadScopeDepth_(0) {
+        localFirstDefOfManySymTable = make_unique<UnorderedMap<core::SymbolRef, core::LocOffsets>>();
+    }
 
-    void preTransformClassDef(core::Context ctx, ast::ExpressionPtr &tree) {
-        nesting_ = make_unique<Nesting>(std::move(nesting_), ast::cast_tree_nonnull<ast::ClassDef>(tree).symbol);
+    void preTransformMethodDef(core::Context ctx, ast::ExpressionPtr &tree) {
+        if (loadScopeDepth_ >= 0) {
+            loadScopeDepth_++;
+        }
+    }
+
+    void postTransformMethodDef(core::Context ctx, ast::ExpressionPtr &tree) {
+        if (loadScopeDepth_ > 0) {
+            loadScopeDepth_--;
+        }
+    }
+
+    void preTransformBlock(core::Context ctx, ast::ExpressionPtr &tree) {
+        if (loadScopeDepth_ >= 0) {
+            loadScopeDepth_++;
+        }
+    }
+
+    void postTransformBlock(core::Context ctx, ast::ExpressionPtr &tree) {
+        if (loadScopeDepth_ > 0) {
+            loadScopeDepth_--;
+        }
     }
 
     void postTransformUnresolvedConstantLit(core::Context ctx, ast::ExpressionPtr &tree) {
         walkUnresolvedConstantLit(ctx, tree);
     }
 
+    void preTransformClassDef(core::Context ctx, ast::ExpressionPtr &tree) {
+        auto &original = ast::cast_tree_nonnull<ast::ClassDef>(tree);
+        auto sym = original.symbol;
+
+        // Populate local first definitions table for out-of-order reference checking.
+        // This needs to happen *iff* the symbol is:
+        //   a). defined only in this file and nowhere else
+        //   b). defined multiple times in this file
```

Comment:
This condition for _populating_ the table seems like it's hard to satisfy, given that we don't store multiple locs per file.  So how do we tell that there are multiple definitions?

And in any event, the error added in this PR isn't checking for multiple definitions, it's checking for the relationship between a definition and a reference, which is not the same thing.  It seems like the comment here (and the similar one for assignments, below) ought to be adjusted?

Reply:
We know that there are multiple by seeing if the current loc is not the same as the canonical loc. That can only happen if there are multiple definitions, afaik.

While this PR isn't directly checking for multiple defs, we need to know the first def (if there are multiple) so we can adjust for pre-declarations. Can change the comment to make that clear.

Reply:
OK, I think I was just confused when I wrote this comment, thinking that the conditions listed here applied to the conditional just below, when in reality, two of these conditions apply to the inner condition a few lines later.

Reply:
Yes, b) is only for perf.